src/main/java/echoserver/EchoServer.java

```diff
@@ -3,10 +3,18 @@
 import java.io.IOException;
 
 public class EchoServer {
-    public void echo(Connection connection) throws IOException {
-        String input;
-        while ((input = connection.read()) != null) {
-            connection.write(input);
+    private final Openable listener;
+    private final Echoable echoClient;
+
+    public EchoServer(Openable listener, Echoable echoClient) {
+        this.listener = listener;
+        this.echoClient = echoClient;
```

Comment:
Any reason you decided to name these `listener` and `echoClient`?

Reply:
Just matching the class names.

Reply:
I see. They seem to map to concrete types, but here you are depending on interfaces. Would it make more sense to have names that map to the interface? What if you passed in an `Openable` that was not a `Listener`?

Reply:
good point

Reply:
Although I like listener as it listens for a connection

Reply:
👍 In the end, there isn't a functional difference, it is more about telling a good story to the reader and using [language that is clear that you and the stakeholders agree on](https://www.martinfowler.com/bliki/UbiquitousLanguage.html).

If you like `listener` more then it sounds like maybe the abstraction (i.e. interface) has the wrong name.

No change needed here, it just felt like a good opportunity to getting you thinking about it.